Identity core

scrawfor99 opened the pull request:
<!--  Thanks for sending a pull request, here are some tips:

1. If this is a fix for an undisclosed security vulnerability, please STOP. All security vulnerability reporting and fixes should be done as per our security policy https://github.com/opensearch-project/OpenSearch/security/policy
2. If this is your first time, please read our contributor guidelines: https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md and developer guide https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md
3. Ensure you have added or ran the appropriate tests for your PR: https://github.com/opensearch-project/OpenSearch/blob/main/TESTING.md
-->

### Description
[Describe what this change achieves]

### Issues Resolved
[List any issues this PR will resolve]

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

scrawfor99 commented:
Hi @dblock, this should be ready for a further review. I am hoping to get this merged as soon as possible to unblock further work in core. Let me know what else you would like done. I addressed the comments left on the previous PR other than the addition of a `@NotNull` annotation to the interface. Adding that annotation was causing the connection to hang and failing the BasicAuth tests.

scrawfor99 commented:
Hi @dblock, @saratvemulapalli, @reta, and @andrross, I know you are all busy but if any of you had the availability today to provide some reviews so we could move forward with some of the Security support for Extensions that would be greatly appreciated. If you don't have time today, I totally understand but figured I would tag you as it looked like you had all interacted with Peter's PR previously. Thank you!

reta commented on CHANGELOG.md:
```diff
@@ -12,20 +12,37 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 
 ### Dependencies
 - Bump `log4j-core` from 2.18.0 to 2.19.0
+- Bumps `reactor-netty-http` from 1.0.24 to 1.1.2
```
The changelog is a bit messed up :(

scrawfor99 replied:
Ooops must have messed up with the rebase on my end.

reta commented on modules/identity-shiro/src/main/java/org/opensearch/identity/shiro/AuthTokenHandler.java:
```diff
@@ -0,0 +1,41 @@
+ */
+class AuthTokenHandler {
+
+    private static final Logger logger = LogManager.getLogger(AuthTokenHandler.class);
```
Probably could be removed, it is not used

scrawfor99 replied:
Good call, I will take it out.

reta commented on server/src/main/java/org/opensearch/identity/NativePrincipal.java:
```diff
@@ -0,0 +1,35 @@
+ *
+ * @opensearch.experimental
+ */
+public class NativePrincipal {
```
`InternalPrincipal` may be? Do we actually need it? `NamedPrincipal` looks to be sufficient ...

scrawfor99 replied:
InternalPrincipal sounds good to me. I could remove this in favor of only using `NamedPrincipal` with a few caveats. Basically, the `InternalPrincipal` (here NativePrincipal), is used for handling the NoopSubject class. This class is in turn used for handling scenarios where the IdentityPlugin does not enforce any authc/z requirements. 

The reason the `NamedPrincipal` is not used for both scenarios is because the `NamedPrincipal` implements the Java `Principal` interface which expects things like `hashCode()` etc. to be overwritten. I could of course makes these null in the Noop scenario if that is preferred.

What sounds best to you?

reta replied:
> What sounds best to you?

Your suggestion to remove it sounds good - I think we could always introduce additional principals later on but it seems like this one is not needed (and to your point, it is not a principal per se - it does not implement `Principal`), thank you

scrawfor99 replied:
I went ahead and moved the internal principal code into the `NamedPrincipal` class as just a special instance that has a static constant.

reta commented on server/src/main/java/org/opensearch/rest/RestController.java:
```diff
@@ -505,6 +518,40 @@ private void handleBadRequest(String uri, RestRequest.Method method, RestChannel
+                );
+                channel.sendResponse(bytesRestResponse);
+            } catch (final Exception _ignored) {
+                // TODO: clean this up
```
We probably want to send the error back to the channel

scrawfor99 replied:
Yeah that makes sense, I will look into adding that here.

reta commented on server/src/main/java/org/opensearch/plugins/IdentityPlugin.java:
```diff
@@ -0,0 +1,24 @@
+    /**
+     * Get the current subject
+     * */
+    public Subject getSubject();
```
Question please, can `getSubject()` return null? If not - could we add Javadocs that subject is required and cannot be null? Thank you.

reta replied:
I think just documenting it would be good at this point, since this is a plugin which could be implemented externally, clarifying exceptions would be fine.

reta commented:
@scrawfor99 I like it, a few minor comments but LGTM otherwise! Apologies for a delay in review

scrawfor99 commented:
> @scrawfor99 I like it, a few minor comments but LGTM otherwise! Apologies for a delay in review

Hi @reta, don't apologize! I appreciate all your comments and will go ahead and address them now. Apologies about pinging you only for the tests to now fail on a text change but such is life... I will address your comments and try to correct the test issue now. Thank you again :)

reta commented on server/src/main/java/org/opensearch/rest/RestController.java:
```diff
@@ -505,6 +518,41 @@ private void handleBadRequest(String uri, RestRequest.Method method, RestChannel
+                    e.getMessage()
+                );
+                channel.sendResponse(bytesRestResponse);
+            } catch (final Exception _ignored) {
```
Nit, not ignored anymore :-)

```suggestion
            } catch (final Exception ex) {
```

reta commented on CHANGELOG.md:
```diff
@@ -12,20 +12,37 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 
 ### Dependencies
 - Bump `log4j-core` from 2.18.0 to 2.19.0
+- Bump `reactor-netty-http` from 1.0.24 to 1.1.2
```
The CHANGELOG is still problematic ...

scrawfor99 replied:
Let me try that again.

cwperks commented:
@scrawfor99 What do you think about creating a PR that only introduces the next extension point `IdentityPlugin` with a companion PR that shows how it will be implemented in the Security plugin for extensions use-cases? I put a PR together against my own fork with only `IdentityPlugin` related changes - no `identity-shiro` module: https://github.com/cwperks/OpenSearch/pull/56

scrawfor99 commented:
Hi @cwperks, this PR is picking up where Peter's PR left off a couple weeks ago. This likely would have already been merged had Peter been the one driving this instead of myself. I am doing my best to take over where he left off however and would prefer to not scrap it for an alternative unless requested by the repository maintainers. That being said, I am happy to review your demo and perhaps you would be interested in opening your own PR against this branch or against core as a follow-up.

peternied commented:
Thanks for jumping on this @scrawfor99 I'll take a look!

scrawfor99 commented:
Hi @peternied,

Thank you for offering to take a look.

Here are the differences from your original PR so you can be aware:

1. Updated to match core main.
2. Added noop IdentityPlugin 
3. Removed the idea of a noop principal and just made this a static case of the `NamedPrincipal` 
4. Added documentation saying that `getPrincipal` should never be null 
5. Fixed old tests that had broken from changes 
6. Removed the `IdentitySmokeTest` class and related tests. They were causing the cluster to hang on the basic auth tests. I took them out and added the configuration setup into the regular SmokeTest location under `/http/` and that works now.
7. Renamed a couple things per other's requests and added a couple error messages.

peternied commented on CHANGELOG.md:
```diff
@@ -12,28 +12,30 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 
```
I think most of these changes should be reverted

scrawfor99 replied:
So for this, there was a ton of issues with resolving the merge conflicts on the CHANGELOG. There were repeats of a lot of stuff. I ended up having to manually comb through and make sure we only kept the change to the most recent version. I then reordered the changes to be in order of PR number while I was doing it. 

If you let me know which changes you want swapped back, I can check to see if I messed something up.

scrawfor99 replied:
For instance `azure-storage-common` was listed as being bumped from 12.18.0 to 12.18.1 but also from 12.18.1 to 12.19.3. So I just kept the latter since it didn't make sense to list a middle version. I will go through and double check each individually again and swap cases to show the earliest to the latest.

peternied replied:
I think you can just revert this whole file and only add in the line for these changes

scrawfor99 replied:
Sounds good

peternied commented on gradle/missing-javadoc.gradle:
```diff
@@ -107,6 +107,7 @@ configure([
   project(":modules:aggs-matrix-stats"),
   project(":modules:analysis-common"),
   project(":modules:geo"),
+  project(":modules:identity-shiro"),
```
Can we remove this exclusion?

scrawfor99 replied:
Yeah no problem.

scrawfor99 replied:
This will have to stay in. I added javadocs everywhere but the compiler still complains about missing javadocs even when there are simply implied default constructors. This does not make much sense but it does not look like we are currently able to easily configure these settings. I may have missed it etc. but from what I saw, we would need to swap the settings for everything to change the strict-ness level.

peternied commented on qa/smoke-test-http/src/test/java/org/opensearch/http/IdentityAuthenticationIT.java:
```diff
@@ -0,0 +1,88 @@
+import static org.hamcrest.core.StringContains.containsString;
+
+@OpenSearchIntegTestCase.ClusterScope(scope = OpenSearchIntegTestCase.Scope.SUITE, supportsDedicatedMasters = false, numDataNodes = 1)
+public class IdentityAuthenticationIT extends HttpSmokeTestCase {
```
Is there a way to have this test inside the identity-shiro module?

scrawfor99 replied:
Ehhh, it is not looking good... Basically, if you have this as part of the module the node will not close after the tests execute. So you will successful auth but then the nodes stay open and the test fails as part of the closing process. I tried narrowing down the cause using a debugger and stepping through but it is incredibly complicated to follow the steps that the node takes when executing the requests.

Is there a reason we cannot/should not keep this in the normal test suite? I moved it here since you can add whatever settings you want as part of the Setup step for these tests so I was able to get the tests to work correctly this way.

scrawfor99 replied:
The cluster continues to hang when we try to separate out the basic auth tests into the module itself. Is there an alternative to moving the basic auth tests into the module that will work?

peternied commented on server/src/main/java/org/opensearch/common/util/FeatureFlags.java:
```diff
@@ -91,4 +96,6 @@ public static boolean isEnabled(String featureFlagName) {
 
     public static final Setting<Boolean> SEARCH_PIPELINE_SETTING = Setting.boolSetting(SEARCH_PIPELINE, false, Property.NodeScope);
+
+    public static final Setting<Boolean> IDENTITY_SETTING = Setting.boolSetting(IDENTITY, false, Property.NodeScope);
```
I think we should remove this high level flag, what do you think?

My rational to remove, either an Identity plugin is installed or it isn't, when this feature is no longer experimental doesn't seem very clear.  If there are specific features I would rather have more traditional configuration on/off around them instead of this broader area.

scrawfor99 replied:
Yeah that sounds good to me.

peternied commented on server/src/test/java/org/opensearch/rest/RestControllerTests.java:
```diff
@@ -52,6 +52,7 @@
 import org.opensearch.http.HttpResponse;
 import org.opensearch.http.HttpServerTransport;
 import org.opensearch.http.HttpStats;
+import org.opensearch.identity.IdentityService;
```
Could you add new unit test methods in this class to verify authentication is attempted, it can succeed, if there is an authentication failure the request is rejected?

scrawfor99 replied:
Sure let me add some.

scrawfor99 replied:
So for the RestController it does not seem like there an easy way of adding tests for checking this. 

After reviewing the existing code and tests I came up with something along these lines:

```

  public void testRestAuthenticationSucceeds() {
        Map<String, List<String>> restHeaders = new HashMap<>();
        restHeaders.put("Authorization", Collections.singletonList("Basic YWRtaW46YWRtaW4="));
        final FakeRestRequest fakeRestRequest = new FakeRestRequest.Builder(NamedXContentRegistry.EMPTY).withMethod(RestRequest.Method.GET)
            .withPath("/favicon.ico")
            .withHeaders(restHeaders)
            .build();
        final AssertingChannel channel = new AssertingChannel(fakeRestRequest, false, RestStatus.OK);
        restController.dispatchRequest(fakeRestRequest, channel, client.threadPool().getThreadContext());
        assertTrue(channel.getSendResponseCalled());
        assertThat(channel.getRestResponse().contentType(), containsString("image/x-icon"));
    }

    public void testRestAuthenticationFails() {
        Map<String, List<String>> restHeaders = new HashMap<>();
        restHeaders.put("Authorization", Collections.singletonList("Basic sdfjalksdjf;a"));
        final FakeRestRequest fakeRestRequest = new FakeRestRequest.Builder(NamedXContentRegistry.EMPTY).withMethod(RestRequest.Method.GET)
            .withPath("/favicon.ico")
            .withHeaders(restHeaders)
            .build();
        final AssertingChannel channel = new AssertingChannel(fakeRestRequest, false, RestStatus.BAD_REQUEST);
        restController.dispatchRequest(fakeRestRequest, channel, client.threadPool().getThreadContext());
        assertTrue(channel.getSendResponseCalled());
        assertFalse(channel.getRestResponse().contentType().contains("image/x-icon"));
    }
    
```

Unfortunately, both tests result in a 200. From what I have determined, there is not any built-in authentication process happening at the RestController level. This makes sense considering the plugin architecture and the way that Security is used as a wrapper for Rest Requests. I am looking into if there are any analogous tests in the Security plugin that may help with setting up what you are asking but I won't have these added by EOD.

peternied replied:
Want to make an issue for this to follow up later?

scrawfor99 replied:
Yeah I will do that now

peternied commented on server/src/main/java/org/opensearch/identity/IdentityService.java:
```diff
@@ -0,0 +1,52 @@
+    public IdentityService(final Settings settings, final List<IdentityPlugin> identityPlugins) {
+        this.settings = settings;
+
+        if (identityPlugins.size() == 0) {
```
Nit; consider adding unit tests to verify the behavior of this class when there are 0, 1, or many identity plugins

scrawfor99 replied:
Yeah, sounds good. I will take a look at adding those alongside the other ones.

peternied reviewed (approved):
Overall looks great, minor comments.

scrawfor99 commented on qa/smoke-test-http/build.gradle:
```diff
@@ -35,7 +35,9 @@ apply plugin: 'opensearch.test-with-dependencies'
 dependencies {
   testImplementation project(path: ':modules:transport-netty4') // for http
-  testImplementation project(path: ':plugins:transport-nio') // for http
+  testImplementation project(path: ':plugins:transport-nio')
```
Yeah will do.

scrawfor99 commented on server/src/main/java/org/opensearch/identity/NamedPrincipal.java:
```diff
@@ -0,0 +1,65 @@
+        return "NamedPrincipal(" + "name=" + name + ")";
+    }
+
+    /**
+     * Returns the NamedPrincipal instanced, should only be used for UNAUTHENTICATED
+     *
+     * @return The specific principal instance
+     */
+    public Principal getPrincipal() {
+        return this;
+    }
+
```
Will do

cwperks commented on server/src/main/java/org/opensearch/identity/Subject.java:
```diff
@@ -0,0 +1,32 @@
+     * throws SubjectNotFound
+     * throws SubjectDisabled
+     */
+    void login(final AuthToken token);
```
Do you think that `login` may imply that a session is created? Maybe `authenticateRequest`?

Do you think it would make any sense to introduce `boolean isAuthenticated();` to this interface for a convenient function to check if the current subject is authenticated?

scrawfor99 replied:
Sure I can switch the name to `authenticateRequest` 

I will look at the code and see if the check method makes sense too!

scrawfor99 commented on server/src/main/java/org/opensearch/identity/IdentityService.java:
```diff
@@ -0,0 +1,52 @@
+        }
+
+        logger.info("Identity module loaded with " + identityPlugin.getClass().getName());
+        logger.info("Current subject " + getSubject());
```
Will do

scrawfor99 commented on server/src/main/java/org/opensearch/rest/RestController.java:
```diff
@@ -505,6 +518,41 @@ private void handleBadRequest(String uri, RestRequest.Method method, RestChannel
+            }
+            final Subject currentSubject = identityService.getSubject();
+            currentSubject.login(token);
+            logger.info("Logged in as user " + currentSubject);
```
No problem

cwperks commented:
@scrawfor99 By introducing a native module and enforcing that only 1 `IdentityPlugin` can be installed at any time, how does a user install a custom plugin? With this PR does it default to using the new `identity-shiro` module being introduced or does it pick up the `NoopIdentityPlugin`?

cwperks commented on server/src/main/java/org/opensearch/identity/IdentityService.java:
```diff
@@ -0,0 +1,49 @@
+            identityPlugin = identityPlugins.get(0);
+        } else {
+            throw new OpenSearchException(
+                "Multiple identity plugins are not supported, found: "
```
Since this PR introduces a default module and this file enforces at most 1 `IdentityPlugin` is installed at a time there is no way to install an IdentityPlugin with this. The opensearch node fails with this error:

```
OpenSearchException[Multiple identity plugins are not supported, found: org.opensearch.identity.shiro.ShiroIdentityPlugin,org.opensearch.security.OpenSearchSecurityPlugin]
	at org.opensearch.identity.IdentityService.<init>(IdentityService.java:38)
	at org.opensearch.node.Node.<init>(Node.java:464)
	at org.opensearch.node.Node.<init>(Node.java:377)
	at org.opensearch.bootstrap.Bootstrap$5.<init>(Bootstrap.java:242)
	at org.opensearch.bootstrap.Bootstrap.setup(Bootstrap.java:242)
	at org.opensearch.bootstrap.Bootstrap.init(Bootstrap.java:404)
	at org.opensearch.bootstrap.OpenSearch.init(OpenSearch.java:180)
	at org.opensearch.bootstrap.OpenSearch.execute(OpenSearch.java:171)
	at org.opensearch.cli.EnvironmentAwareCommand.execute(EnvironmentAwareCommand.java:104)
	at org.opensearch.cli.Command.mainWithoutErrorHandling(Command.java:138)
	at org.opensearch.cli.Command.main(Command.java:101)
	at org.opensearch.bootstrap.OpenSearch.main(OpenSearch.java:137)
	at org.opensearch.bootstrap.OpenSearch.main(OpenSearch.java:103)
For complete error details, refer to the log at /Users/cwperx/Desktop/distributions/opensearch-stephen-identity/opensearch-3.0.0-SNAPSHOT/logs/opensearch.log
```

scrawfor99 replied:
Hi @cwperks, great catch. This can be resolved in a couple minor changes--making the identity plugin used configurable or allowing shiro to be turned off--but it may be best to wait until @peternied is available and ask him what he thinks considering much of this was his original work.

peternied replied:
Since moving forward the we are not intending on using [1] security plugin build out of core, I think we shouldn't make it harder for identity plugin developers to get started by requiring them to remove/uninstall the identity-shiro plugin.  I think we can revisit how this works as we get feedback from other security implementations, until then I think we should default to use the 'no op' provider.

To support this seems like we might need to move the location from module to plugin.  @scrawfor99 I think your question [2] framed this well and we can figure out how to best move forward with a recommendation from the maintainers.
 
- [1] https://github.com/opensearch-project/security/issues/2602
- [2] https://github.com/opensearch-project/OpenSearch/pull/7246#issuecomment-1527990034

scrawfor99 commented:
Hi @reta, @dblock, and @saratvemulapalli. I am tagging you because we are trying to move this PR forward but wanted your input. As mentioned by @cwperks, in this [comment](https://github.com/opensearch-project/OpenSearch/pull/7246#discussion_r1179681473), the way that this PR is configured will need to be modified so that the IdentityPlugin being used is configurable. Currently, we cannot install any other IdentityPlugins since Shiro will always be installed by virtue of being a module.

Before I change this, I wanted to get the opinion of some core maintainers about whether we could move forward with this PR afterwards. An alternative approach to getting our changes into Core would be to start with a barebones Identity Interface and then merging smaller changes. This may be easier to review, but will not show the same functional demonstration that the Shiro implementation @peternied setup would. 

Which direction do you prefer I take?

reta commented:
> As mentioned by @cwperks, in this [comment](https://github.com/opensearch-project/OpenSearch/pull/7246#discussion_r1179681473), the way that this PR is configured will need to be modified so that the IdentityPlugin being used is configurable.

I think this is a good point but do we want it to be configurable? I mean if in 99.9% use cases the default is enough, the other 0.1% of the users could always uninstall it and replace with the custom plugin. Also, do we always want to bundle it? May be we could just move it to `plugins` folder to become opt-in plugin?

scrawfor99 commented:
> 

Hi @reta, I think moving it from modules to a bundled plugin would be an appropriate solution. The reason we are planning on bundling it would be so that users could test the functionality of the Identity Interface and potentially create their own implementations. Basically, we are going a similar route as the SDK's HelloWorld extension where we doubt anyone will choose to use Shiro but this lets us demonstrate how the interface works. 

Moving the code from a module to a plugin should allow us to ignore the installation issue and we could look into further configuration changes as needed.

reta commented on server/src/main/java/org/opensearch/node/Node.java:
```diff
@@ -454,6 +456,15 @@ protected Node(
+                logger.info("Identity on so found plugins implementing: " + pluginsService.filterPlugins(IdentityPlugin.class).toString());
+                identityPlugins.addAll(pluginsService.filterPlugins(IdentityPlugin.class));
+            }
+            logger.info("Creating identity service in Node.java with identity plugins: " + identityPlugins.toString());
```
Oops, @scrawfor99 my apologies, I missed that, probably `Node.java` is too precise:

```suggestion
            logger.info("Creating identity service for node with identity plugins: " + identityPlugins.toString());
```

scrawfor99 replied:
Oops, I did not mean to include that print. I will remove it.

peternied commented on server/src/main/java/org/opensearch/rest/RestController.java:
```diff
@@ -395,6 +403,11 @@ private void tryAllHandlers(final RestRequest request, final RestChannel channel
                     }
                 } else {
+                    if (FeatureFlags.isEnabled(FeatureFlags.IDENTITY)) {
+                        if (!handleAuthenticateUser(request, channel)) {
```
Yes, I imagine at some point the Rest Wrapper would be completely removed from the security plugin as there will be tighter interfaces that better support audit logging or www-authentication features.
